command-parser/src/config.rs

```diff
@@ -120,6 +140,105 @@ impl<'a> Config<'a> {
     }
 }
 
+/// A builder struct for building commands.
+///
+/// The [`add`] method needs to be called for the command to be added to the  [`Config`].
+///
+/// Commands are not case sensitive by default. Use [`case_sensitive`] to enable this.
+///
+/// # Examples
+///
+/// ```rust
+/// use dawn_command_parser::Config;
+///
+/// let mut config = Config::new();
+/// // Adds a case sensitive command to the config.
+/// config.command("ping").case_sensitive().add();
+/// assert_eq!(1, config.commands().len());
+/// ```
+///
+/// [`add`]: #method.add
+/// [`case_sensitive`]: #method.case_sensitive
+/// [`Config`]: struct.Config.html
+pub struct CommandBuilder<'a, 'b> {
+    name: String,
+    case_sensitive: bool,
+    config: &'a mut Config<'b>,
+}
+
+impl<'a, 'b> CommandBuilder<'a, 'b> {
+    /// Adds the command to the [`Config`].
+    ///
+    /// [`Config`]: struct.Config.html
+    pub fn add(self) {
+        self.config.add_command(self.name, self.case_sensitive);
+    }
+
+    /// Makes the command only match if the case is the same.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use dawn_command_parser::{Config, Parser};
+    ///
+    /// let mut config = Config::new();
+    /// config.add_prefix("!");
+    /// config.command("ping").case_sensitive().add();
+    /// let parser = Parser::new(config);
+    /// assert!(parser.parse("!ping should work").is_some());
+    /// assert!(parser.parse("!PiNg shouldn't work").is_none());
+    /// ```
+    pub fn case_sensitive(mut self) -> Self {
+        self.case_sensitive = true;
+        self
+    }
+
+    /// Makes the command match regardless of case.
+    /// This is set by default.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use dawn_command_parser::{Config, Parser};
+    ///
+    /// let mut config = Config::new();
+    /// config.add_prefix("!");
+    /// config.command("ping").case_insensitive().add();
+    /// let parser = Parser::new(config);
+    /// assert!(parser.parse("!ping should work").is_some());
+    /// assert!(parser.parse("!PiNg should also work").is_some());
+    /// ```
+    pub fn case_insensitive(mut self) -> Self {
+        self.case_sensitive = false;
+        self
+    }
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub enum CaseSensitivity {
+    Insensitive(UniCase<String>),
+    Sensitive(String),
+}
+
+impl AsRef<str> for CaseSensitivity {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::Insensitive(u) => u.as_str(),
+            Self::Sensitive(s) => s.as_str(),
+        }
+    }
+}
+
+impl PartialEq<str> for CaseSensitivity {
+    fn eq(&self, other: &str) -> bool {
+        match self {
+            Self::Insensitive(u) if u == &UniCase::new(other) => true,
+            Self::Sensitive(s) if s.eq(other) => true,
+            _ => false,
```

Comment:
Would it not be more simple to do something like:
```rust
            Self::Insensitive(u) => u == &UniCase::new(other),
            Self::Sensitive(s) => s.eq(other),
```

Reply:
Yup, will also change the `eq` with just a normal equality operator ( `==` ).